### PR TITLE
Cleaned up ts MarginAccount and MarginPool

### PR DESCRIPTION
**Pull request by Tristyn**
Derive pubkeys upfront
Accept `Address` params which are `PublicKey | string`
Reduced size of constructors using config
Anchor methods api
Wallet adapter api
@solana/buffer-layout for better types
Added .prettierrc formatting
Added strict tsconfig but haven't turned it on yet

Haven't ran tests yet so it's a draft

### CODE REVIEW

**Comment by jetjake** [comment]
Test results from anchor test which i think are expected: 
  
  margin account
    ✔ Fund payer (1119ms)
    ✔ Create two user wallets (1009ms)
    1) Create authority
    ✔ Create margin accounts (868ms)

  margin account
    ✔ Fund payer (474ms)
    ✔ Create tokens (970ms)
    ✔ Create oracles (985ms)
    2) Create authority
    3) Register adapter
    4) Create margin pools
    ✔ Create our two user wallets, with some SOL funding to get started (841ms)
    ✔ Initialize the margin accounts for each user (952ms)
    ✔ Create some tokens for each user to deposit (1914ms)
    ✔ Set the prices for each token (995ms)
    5) Deposit user funds into their margin accounts
    ✔ Set the prices for each token (906ms)
    6) Have each user borrow the other's funds
    7) Users repay their loans
    8) Users withdraw their funds
    ✔ Now verify that the users got all their tokens back

  pyth-oracle
    ✔ Fund payer (407ms)
    ✔ initialize (509ms)
    ✔ change feed price (944ms)

  margin swap
    ✔ Fund payer (927ms)
    ✔ Create token swap pool (1441ms)
    ✔ create and mint tokens (2854ms)
    9) createTokenSwap (constant product)
    10) deposit all token types
    11) withdraw all token types
    12) swap
    13) create account, approve, swap all at once
    14) deposit one exact amount in
    15) withrdaw one exact amount out


  18 passing (22s)
  15 failing

  1) margin account
       Create authority:
     Error: failed to send transaction: Transaction simulation failed: Error processing Instruction 0: custom program error: 0x0
      at Connection.sendEncodedTransaction (node_modules/@solana/web3.js/src/connection.ts:4153:13)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at Connection.sendRawTransaction (node_modules/@solana/web3.js/src/connection.ts:4115:20)
      at Connection.sendTransaction (node_modules/@solana/web3.js/src/connection.ts:4103:12)

  2) margin account
       Create authority:
     Error: failed to send transaction: Transaction simulation failed: Error processing Instruction 0: custom program error: 0x0
      at Connection.sendEncodedTransaction (node_modules/@solana/web3.js/src/connection.ts:4153:13)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at Connection.sendRawTransaction (node_modules/@solana/web3.js/src/connection.ts:4115:20)
      at Connection.sendTransaction (node_modules/@solana/web3.js/src/connection.ts:4103:12)

  3) margin account
       Register adapter:
     Error: failed to send transaction: Transaction simulation failed: Error processing Instruction 0: custom program error: 0x7dc
      at Connection.sendEncodedTransaction (node_modules/@solana/web3.js/src/connection.ts:4153:13)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at Connection.sendRawTransaction (node_modules/@solana/web3.js/src/connection.ts:4115:20)
      at Connection.sendTransaction (node_modules/@solana/web3.js/src/connection.ts:4103:12)

  4) margin account
       Create margin pools:
     Error: failed to send transaction: Transaction simulation failed: Error processing Instruction 0: Cross-program invocation with unauthorized signer or writable account
      at Connection.sendEncodedTransaction (node_modules/@solana/web3.js/src/connection.ts:4153:13)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at Connection.sendRawTransaction (node_modules/@solana/web3.js/src/connection.ts:4115:20)
      at sendAndConfirmRawTransaction (node_modules/@project-serum/anchor/src/provider.ts:284:21)
      at AnchorProvider.sendAndConfirm (node_modules/@project-serum/anchor/src/provider.ts:144:14)

  5) margin account
       Deposit user funds into their margin accounts:
     Error: failed to send transaction: Transaction simulation failed: Error processing Instruction 0: custom program error: 0xbc4
      at Connection.sendEncodedTransaction (node_modules/@solana/web3.js/src/connection.ts:4153:13)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at Connection.sendRawTransaction (node_modules/@solana/web3.js/src/connection.ts:4115:20)
      at sendAndConfirmRawTransaction (node_modules/@project-serum/anchor/src/provider.ts:284:21)
      at AnchorProvider.sendAndConfirm (node_modules/@project-serum/anchor/src/provider.ts:144:14)

  6) margin account
       Have each user borrow the other's funds:
     TypeError: Cannot read property 'marginBorrow' of undefined
      at /Users/jakemozack/Development/jet/jet-v2/tests/integration/pool/marginPool.test.ts:211:26
      at Generator.next (<anonymous>)
      at /Users/jakemozack/Development/jet/jet-v2/tests/integration/pool/marginPool.test.ts:31:71
      at new Promise (<anonymous>)
      at __awaiter (tests/integration/pool/marginPool.test.ts:27:12)
      at Context.<anonymous> (tests/integration/pool/marginPool.test.ts:210:60)
      at processImmediate (node:internal/timers:464:21)

  7) margin account
       Users repay their loans:
     TypeError: Cannot read property 'marginRepay' of undefined
      at /Users/jakemozack/Development/jet/jet-v2/tests/integration/pool/marginPool.test.ts:216:26
      at Generator.next (<anonymous>)
      at /Users/jakemozack/Development/jet/jet-v2/tests/integration/pool/marginPool.test.ts:31:71
      at new Promise (<anonymous>)
      at __awaiter (tests/integration/pool/marginPool.test.ts:27:12)
      at Context.<anonymous> (tests/integration/pool/marginPool.test.ts:215:44)
      at processImmediate (node:internal/timers:464:21)

  8) margin account
       Users withdraw their funds:
     Error: failed to send transaction: Transaction simulation failed: Error processing Instruction 0: custom program error: 0xbc4
      at Connection.sendEncodedTransaction (node_modules/@solana/web3.js/src/connection.ts:4153:13)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at Connection.sendRawTransaction (node_modules/@solana/web3.js/src/connection.ts:4115:20)
      at sendAndConfirmRawTransaction (node_modules/@project-serum/anchor/src/provider.ts:284:21)
      at AnchorProvider.sendAndConfirm (node_modules/@project-serum/anchor/src/provider.ts:144:14)

  9) margin swap
       createTokenSwap (constant product):
     Error: failed to send transaction: Transaction simulation failed: Attempt to load a program that does not exist
      at Connection.sendEncodedTransaction (node_modules/@solana/web3.js/src/connection.ts:4153:13)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at Connection.sendRawTransaction (node_modules/@solana/web3.js/src/connection.ts:4115:20)
      at Connection.sendTransaction (node_modules/@solana/web3.js/src/connection.ts:4103:12)
      at sendAndConfirmTransaction (node_modules/@solana/web3.js/src/util/send-and-confirm-transaction.ts:30:21)

  10) margin swap
       deposit all token types:
     TypeError: Cannot read property 'tokenSwap' of undefined
      at /Users/jakemozack/Development/jet/jet-v2/tests/integration/swap/marginSwap.test.ts:218:78
      at Generator.next (<anonymous>)
      at /Users/jakemozack/Development/jet/jet-v2/tests/integration/swap/marginSwap.test.ts:31:71
      at new Promise (<anonymous>)
      at __awaiter (tests/integration/swap/marginSwap.test.ts:27:12)
      at Context.<anonymous> (tests/integration/swap/marginSwap.test.ts:217:44)
      at processImmediate (node:internal/timers:464:21)

  11) margin swap
       withdraw all token types:
     TypeError: Cannot read property 'tokenSwap' of undefined
      at /Users/jakemozack/Development/jet/jet-v2/tests/integration/swap/marginSwap.test.ts:307:78
      at Generator.next (<anonymous>)
      at /Users/jakemozack/Development/jet/jet-v2/tests/integration/swap/marginSwap.test.ts:31:71
      at new Promise (<anonymous>)
      at __awaiter (tests/integration/swap/marginSwap.test.ts:27:12)
      at Context.<anonymous> (tests/integration/swap/marginSwap.test.ts:306:45)
      at processImmediate (node:internal/timers:464:21)

  12) margin swap
       swap:
     TypeError: Cannot read property 'tokenSwap' of undefined
      at /Users/jakemozack/Development/jet/jet-v2/tests/integration/swap/marginSwap.test.ts:408:22
      at Generator.next (<anonymous>)
      at fulfilled (tests/integration/swap/marginSwap.test.ts:28:58)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)

  13) margin swap
       create account, approve, swap all at once:
     TypeError: Cannot read property 'createInitAccountInstruction' of undefined
      at /Users/jakemozack/Development/jet/jet-v2/tests/integration/swap/marginSwap.test.ts:478:18
      at Generator.next (<anonymous>)
      at fulfilled (tests/integration/swap/marginSwap.test.ts:28:58)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)

  14) margin swap
       deposit one exact amount in:
     TypeError: Cannot read property 'tokenSwap' of undefined
      at /Users/jakemozack/Development/jet/jet-v2/tests/integration/swap/marginSwap.test.ts:540:78
      at Generator.next (<anonymous>)
      at /Users/jakemozack/Development/jet/jet-v2/tests/integration/swap/marginSwap.test.ts:31:71
      at new Promise (<anonymous>)
      at __awaiter (tests/integration/swap/marginSwap.test.ts:27:12)
      at Context.<anonymous> (tests/integration/swap/marginSwap.test.ts:536:48)
      at processImmediate (node:internal/timers:464:21)

  15) margin swap
       withrdaw one exact amount out:
     TypeError: Cannot read property 'tokenSwap' of undefined
      at /Users/jakemozack/Development/jet/jet-v2/tests/integration/swap/marginSwap.test.ts:647:78
      at Generator.next (<anonymous>)
      at /Users/jakemozack/Development/jet/jet-v2/tests/integration/swap/marginSwap.test.ts:31:71
      at new Promise (<anonymous>)
      at __awaiter (tests/integration/swap/marginSwap.test.ts:27:12)
      at Context.<anonymous> (tests/integration/swap/marginSwap.test.ts:642:50)
      at processImmediate (node:internal/timers:464:21)

File: libraries/ts/src/margin/marginAccount.ts [hunk 1]
```diff
@@ -1,155 +1,128 @@
...
-    connection: Connection,
-    seed: number,
-  ) {
-    const tx = new Transaction();
-    tx.add(
-      await this.makeCreateAccountInstruction(seed),
-    );
-    return await this.sendTransaction(connection, tx, [this.owner, this.payer]);
+  async refresh() {
+    this.info = await this.programs.margin.account.marginAccount.fetchNullable(this.address)
```
Comment (jetjake) [inline libraries/ts/src/margin/marginAccount.ts:95]: question, why do we have to dig all the way into "this.programs.margin.account.marginAccount" isn't this instance of the class the marginAccount?  I know the rest are static methods.

**Review by jetjake — COMMENTED** [review]
Just one question

File: libraries/ts/src/margin/marginAccount.ts (hunk 1 above)
Comment (Tristyn) [reply libraries/ts/src/margin/marginAccount.ts:95]: Thank you for running the tests while my env is messed up 😭 🙏 

I could pull `this.programs.margin` into it's own class variable to shorten it, but it's only used 5 times in the whole class

There's a lot of object composition in the lib but I don't think it's a code smell so far

**Review by jetjake — APPROVED** [review]

**Comment by jetjake** [comment]
Good to merge.

**Pull request merged**